src/App.tsx

```diff
@@ -51,9 +68,45 @@ function generateHue() {
   return Math.floor(Math.random() * 360);
 }
 
+/** 
+ * This appears to still be the best way to download a file while suggesting a filename.
+ * 
+ * According to [mdn](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-download) 
+ * the download attribute gets ignored on URIs that are not either `same-origin` or use the `blob` or `data` schemes.
+ */
+function downloadUri(uri: string, filename: string) {
+  const downloadAnchor = document.createElement("a");
+  console.log(filename);
+  downloadAnchor.download = filename;
+  downloadAnchor.href = uri;
+  downloadAnchor.click();
+}
+
+async function eventToAsync<EventType extends keyof HTMLElementEventMap>(element: HTMLElement, eventType: EventType) {
+  const abortController = new AbortController();
+  const evt = await new Promise<HTMLElementEventMap[EventType]>((resolve) => {
+    element.addEventListener(eventType, (evt) => resolve(evt), {signal: abortController.signal});
+  })
+  abortController.abort();
+  return evt;
+}
+
+/**
+ * This appears to be the best way to upload a file.
+ */
+async function uploadFile() {
+  const uploadInput = document.createElement("input");
+  uploadInput.type = "file";
+  uploadInput.click();
+  await eventToAsync(uploadInput, "change");
```

Comment:
Not sure if the extra helper function is useful here; you could simplify the code a bit by just inlining the `resolve` logic.

Also, `App.tsx` is pretty large, maybe some of this download/upload logic and `useKeyboardCtrlIntercept` hook could be moved into a separate file?

Reply:
Yes, it was more of an experiment if this can be done with a clean API on my side, when pushing this here I also thought it would probably be best if I get rid of the function again.

Yes, App.tsx is gigantic. It was before, its even more with this commit. I also asked about that specific thing in one of my messages above. I will move this into a seperate file, and if thats fine with you also some other aspects If I find it relevant, even if theyre not a part of this MR directly?

This MR was generally not very tested or intended to be final - i tried to express that in one of the comments on this MR -, and I mostly wanted some initial feedback by you, if my general approach is fine.

Reply:
@ekzhang wait tho. I need to either flatten this with a promise, or make this a synchronous function which explicitly returns a Promise. 

But one of these is required if we want to stay away from callback hell. We have user interaction here, the user has to select the file, and these APIs are using callbacks. If we don't manually wrap something here in a Promise we would have to pass a callback to uploadFile (now called getFileUploadWithDialog btw), which breaks with the style you had in other callbacks, where you used async/await.

We have to have ugly somewhere to switch between these.

But I'll push a change that makes all this somewhat better looking.